tx_test.go

```diff
@@ -1009,3 +1010,48 @@ func TestTxStats_Sub(t *testing.T) {
 	assert.Equal(t, int64(10001), diff.GetWrite())
 	assert.Equal(t, 10009*time.Second, diff.GetWriteTime())
 }
+
+// TestTx_TruncateBeforeWrite ensures the file is truncated ahead whether we sync freelist or not.
+func TestTx_TruncateBeforeWrite(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		return
+	}
+	for _, isSyncFreelist := range []bool{false, true} {
+		// Open the database.
+		db, err := bolt.Open(tempfile(), 0666, &bolt.Options{
+			NoFreelistSync: isSyncFreelist,
+		})
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer os.Remove(db.Path())
+
+		initSize := fileSize(db.Path())
+
+		tx, err := db.Begin(true)
+		if err != nil {
+			t.Fatalf("Error starting tx: %v", err)
+		}
+		bucket := []byte("mybucket")
+		var b *bolt.Bucket
+		if b, err = tx.CreateBucket(bucket); err != nil {
+			t.Fatalf("Error creating bucket: %v", err)
+		}
+		// Put a big enough value to ensure new space must be allocated
+		bigvalue := make([]byte, db.Info().PageSize)
+		if err := b.Put([]byte("key"), bigvalue); err != nil {
+			t.Fatalf("Error on put: %v", err)
+		}
+		if err := tx.Commit(); err != nil {
+			t.Fatalf("Error on commit: %v", err)
+		}
+
+		newSize := fileSize(db.Path())
+		if newSize < initSize*2 {
```

Comment:
I'm not sure why this assertions proves that Grow works... 

I think that real prove starts when we cross the `AllocSize` border and discover that small additional allocations puts us on 2*AllocSize.    Here the file might have just grown as write at offset consequence.

Reply:
Yes the way you mentioned is more obvious, but I have no good idea to make up a database to cross the `AllocSize` exactly...

> Here the file might have just grown as write at offset consequence.

The initial database fills 4 pages, and  I put a value of 1 page size, so there would be around 6 pages.  `mmapSize` normalize the datasz to 32kb or multiple of 32kb, and `db.grow` would truncate to datasz, so I judge by at least double initSize, i.e. 8 pages.

Of course if I can make up datasz to `AllocSize`, it's more obvious to distinguish.

Reply:
1. I think you can iteratively write data of size in order of AllocSize/100. 
And I would guess that around 80th write you would observe the file exceeded 2*AllocSize. 
If we do that write, you should never observe a file of Size x:   `AllocSize<x < 2 * AllocaSize`

2. Seems your math works, assuming we are on 4KB pages (and there are OSes with 64KB pages). 
    So minimal version would be to enforce the 4096 page and put in the comment the math that you made here... 
    But if 1. is not extremely slow, I would recommend trying 1.

Reply:
I adopted the 1. for the readibility. The test time in my laptop is around 10 seconds, which I support is acceptable.

PS the 2. even for odd page size like 64KB, I think it still works. As `mmapSize` normalize the datasz to 32kb or multiple of 32KB, for 64KB page size and 6 page, the datasz would be 512KB=32KB*2^4>64KB*6，the initial datasz is 64KB*4=256KB, so it still obey newSize >= initSize*2. Forgive my rigidness : )